NDTensors/src/dense/dense.jl

```diff
@@ -3,73 +3,145 @@
 #
 using LinearAlgebra: BlasFloat
 
-struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
-  data::VecT
-  function Dense{ElT,VecT}(data::AbstractArray) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+struct Dense{ElT,DataT<:AbstractArray} <: TensorStorage{ElT}
+  data::DataT
+  function Dense{ElT,DataT}(data::DataT) where {ElT,DataT<:AbstractArray}
+    return new{ElT,DataT}(data)
   end
+end
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT,DataT}() where {ElT,DataT<:AbstractArray}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  return Dense{ElT,typedDataT}(typedDataT())
+end
 
-  # Special case of Vector
-  function Dense{ElT,VecT}(data::Vector) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(data)
+# Construct from a set of indices
+# This will fail if zero(ElT) is not defined for the ElT
+function Dense{ElT,DataT}(inds::Tuple) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
 
-  # Special case of Array which gets turned into a Vector
-  function Dense{ElT,VecT}(data::Array) where {ElT,VecT<:AbstractArray{ElT}}
-    return new{ElT,VecT}(vec(data))
+  return Dense{ElT,typedDataT}(generic_zeros(typedDataT, dim(inds)))
+end
+
+function Dense{ElT,DataT}(dim::Integer) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only vector-based datatypes are currently supported.")
+    throw(TypeError)
   end
+  return Dense{ElT,typedDataT}(generic_zeros(typedDataT, dim))
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+function Dense{ElT,DataT}(
+  ::UndefInitializer, inds::Tuple
+) where {ElT,DataT<:AbstractArray{<:Any,N}} where {N}
+  typedDataT = set_eltype_if_unspecified(DataT, ElT)
+  if N > 1
+    println("Only Vector-based datatypes are currently supported.")
+    throw(TypeError)
+  end
+  return Dense{ElT,typedDataT}(similar(typedDataT, dim(inds)))
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+function Dense{ElR,DataT}(data::AbstractArray) where {ElR,DataT<:AbstractArray}
+  if eltype(data) != ElR
+    data = similartype(typeof(data), ElR)(data)
+  end
+  return Dense{ElR,DataT}(data)
```

Comment:
It seems like this would be better handled like:
```julia
function Dense{ElR,DataT}(data::AbstractArray) where {ElR,DataT}
  return Dense{ElR,DataT}(convert(DataT, data))
end
```
`if`-statements with type comparisons like that are basically always better handled through dispatch.

Also, for the version you wrote, if `eltype(data) == ElR`, would it lead to an infinite loop?